### PR TITLE
Support policy groups

### PR DESCRIPTION
**Pull request by mtojek**
## What does this PR do?

This PR introduces policy groups.

Package preview: https://github.com/mtojek/package-spec/tree/132-support-policy-groups/test/packages/input_groups 
_Don't focus on the `policy_groups` name, it's a working name for the test package, in reality it will be `aws`, `nginx`, etc._

## Why is it important?

We need to ensure better granularity for policies, e.g. for the AWS integration.

## Checklist

- [x] I have added test packages to [`test/packages`](https://github.com/elastic/package-spec/tree/master/test/packages) that prove my change is effective.
- [x] I have added an entry in [`versions/N/changelog.yml`](https://github.com/elastic/package-spec/blob/master/versions/1/changelog.yml).

## Related issues

- https://github.com/elastic/package-spec/issues/132

### CODE REVIEW

File: test/packages/policy_groups/manifest.yml
Comment (sorantis) [inline test/packages/policy_groups/manifest.yml:106]: `policy_groups` sounds like a good name, but why do we need them to be policy set specific? I was thinking they'd be generic groups that inputs can refer to.

```yaml
policy_groups:
  - name: logs
    title: "Logs" 
  - name: metrics
    title: "Metrics"
....

policy_templates:
  - name: ec2
    title: AWS EC2
    description: Collect logs and metrics from EC2 service
    inputs:
      - type: s3
        title: Collect logs from EC2 service
        description: Collecting EC2 logs using S3 input
        policy_group: logs
        vars: ...
      - type: aws/metrics
        title: Collect metrics from EC2 service
        description: Collecting EC2 metrics using AWS CloudWatch
        policy_group: metrics

```

Comment (sorantis) [reply test/packages/policy_groups/manifest.yml:106]: @mtojek @ruflin @ycombinator @kaiyan-sheng

Comment (mtojek) [reply test/packages/policy_groups/manifest.yml:106]: Sure that's configurable. I wanted to highlight that it's possible to have even 2 different logs groups. Let me adjust it.

Comment (mtojek) [reply test/packages/policy_groups/manifest.yml]: Fixed

Comment (sorantis) [reply test/packages/policy_groups/manifest.yml]: the input should also be updated to reflect the fix :)

**Comment by sorantis** [comment]
I think we're getting close! We also would like each policy set to be part of one or more categories, i.e. if the user browses the Database category, they should be able to see DynamoDb tile too, so it should be possible to define categories for each policy set. E.g.
```yaml
policy_templates:
  - name: dynamodb
    title: AWS DynamoDB
    description: Collect logs and metrics from DynamoDB service
    categories: [AWS, Cloud, Database]
```

**Comment by mtojek** [comment]
> I think we're getting close! We also would like each policy set to be part of one or more categories, i.e. if the user browses the Database category, they should be able to see DynamoDb tile too, so it should be possible to define categories for each policy set. E.g.

I think we can do this. It would require also a slight modification in the package-registry to cover this change. We have a predefined set of categories that are supported: https://github.com/elastic/package-spec/blob/master/versions/1/manifest.spec.yml#L61 (enum).

**Comment by sorantis** [comment]
@mtojek who's curating this list?

**Comment by mtojek** [comment]
Here is a thread in which we considered introducing another category: https://github.com/elastic/package-registry/pull/671#issuecomment-762328923

You can see all people taking part.

**Comment by mtojek** [comment]
I think I've covered the entire proposal with schema, marking it as ready for review.

**Comment by jen-huang** [comment]
Tagging myself to do some exploratory testing and requirements gathering on Kibana side. Thanks for putting the changes in PR, this makes it much easier.

File: test/packages/policy_groups/manifest.yml
Comment (sorantis) [inline test/packages/policy_groups/manifest.yml:103]: we should call it something else, since effectively we're grouping inputs, the name should reflect that.

Comment (sorantis) [reply test/packages/policy_groups/manifest.yml:103]: we're still stuck on aligning names, so apologies in advance if such comments may disrupt this PR. But we're getting there.

Comment (kaiyan-sheng) [reply test/packages/policy_groups/manifest.yml:103]: How about we keep `groups` here?

File: test/packages/policy_groups/manifest.yml
Comment (kaiyan-sheng) [inline test/packages/policy_groups/manifest.yml:81]: We don't have dynamodb logs right now so should we remove this log section here?

**Comment by kaiyan-sheng** [comment]
@mtojek Will we still have a `data_stream` directory at the [top level](https://github.com/mtojek/package-spec/tree/132-support-policy-groups/test/packages/policy_groups) with `docs`, `img` and `manifest.yml`?

File: test/packages/policy_groups/manifest.yml
Comment (ycombinator) [inline test/packages/policy_groups/manifest.yml:33]: @mtojek is this the array from which Kibana only shows the first element today? Or is that something different?

File: test/packages/policy_groups/manifest.yml
Comment (ycombinator) [reply test/packages/policy_groups/manifest.yml:103]: I'm -1 for calling it just `groups` as that's rather generic; I'd prefer to call it something more descriptive and/or specific.  

Maybe `input_groups` per @sorantis [comment](https://github.com/elastic/package-spec/pull/137/files#r579334776)?

Yay, naming 😄.

**Comment by ycombinator** [comment]
> Will we still have a `data_stream` directory at the top level with docs, img and manifest.yml?

I have the same question. I think we still need this otherwise where else would the ingest pipelines, field definitions, etc. for each data stream be stored? @mtojek if you agree, could you add a couple of data stream folders to this PR please so we can see the fuller picture of the package? Thanks!

File: test/packages/policy_groups/manifest.yml
Comment (ycombinator) [reply test/packages/policy_groups/manifest.yml:33]: Or maybe @jen-huang can weigh in on my question since she's already watching this PR. 🙂

**Comment by ycombinator** [comment]
@sorantis Clarifying question: based on the proposed package structure in this PR, which element in the package's `manifest.yml` would you like to see as separate tiles when a user searches for "aws" (besides the "main" AWS tile itself)?

File: test/packages/policy_groups/manifest.yml
Comment (mtojek) [reply test/packages/policy_groups/manifest.yml:81]: This is just a sample spec showing format opportunities and different cases, I suppose we can leave it here :) But you're right, Kaiyan, the AWS manifest in Integrations will have a different form (real configuration).

**Comment by mtojek** [comment]
@kaiyan-sheng @ycombinator Yes, data streams stay as is :) I will add them to the package to get rid of confusion.

File: test/packages/policy_groups/manifest.yml
Comment (kaiyan-sheng) [reply test/packages/policy_groups/manifest.yml:81]: Sounds good to me :) Thanks for using AWS here for the example!

File: test/packages/policy_groups/manifest.yml
Comment (jen-huang) [reply test/packages/policy_groups/manifest.yml:33]: yes, this is what Kibana uses the first element of today

**Comment by jen-huang** [comment]
(I'll be OOO next week so wanted to give this at least a quick look-over.) The changes so far look good from a Kibana perspective, there is work that needs to be done to support multiple `policy_templates` and the new `policy_groups` but I'm not spotting any obvious breaking changes. I assume that there should be no changes to the agent YAML that Kibana generates?

++ on showing data streams here for clarity

File: test/packages/policy_groups/img/aws-dynamodb.svg
Comment (ruflin) [inline test/packages/policy_groups/img/aws-dynamodb.svg:1]: Nit: Do you need the `aws` prefix here? I would assume if not, it would automatically match the group name?

File: test/packages/policy_groups/manifest.yml
Comment (ruflin) [inline test/packages/policy_groups/manifest.yml:8]: @sorantis Could have a group/set its own categories? Is it always a subset of the package categories?

File: test/packages/policy_groups/manifest.yml
Comment (ruflin) [inline test/packages/policy_groups/manifest.yml:82]: I was first thinking if we should have a global way to specify inputs but this here shows nicely that each config of the input has a different title / description as it is specific to this case. So this is great.

File: test/packages/policy_groups/manifest.yml
Comment (ruflin) [inline test/packages/policy_groups/manifest.yml:103]: Ignoring naming on my end for now ;-)

I'm wondering if the groups should be defined as part of each policy_template or global. Having it local would allow to give more descriptive descriptions. At the same time, I don't think we use it today. Global requires much fewer configuration.

An alternative option is to just not have it in the package at first. Only 2-3 groups are supported, required by the spec and hardcoded into Kibana. Like this everyone is forced to use `logs/metrics` as groups.

**Review by ruflin — COMMENTED** [review]
This looks promising. @mtojek Could you extend the example with one case where a single policy_template with a data_stream has more than 1 input agent template? I think there are a few packages which have it.

File: test/packages/policy_groups/img/aws-dynamodb.svg
Comment (mtojek) [reply test/packages/policy_groups/img/aws-dynamodb.svg:1]: This is not a prefix for the `aws` integration, just a "random" string :) I think I'm in favor of consistency with other icon definitions in package manifest where there is no magic around file prefixes/package/data stream - just a simple file reference. It might be not straightforward for users to guess why some icons are skipped/filtered.

File: test/packages/policy_groups/manifest.yml
Comment (mtojek) [reply test/packages/policy_groups/manifest.yml:103]: Actually your both options are related. I agree that global definitions would require less configuration properties and might be easier to setup. The second option is an interesting solution - we can try to kick off the feature with predefined groups and expose only them to users. We need to remember that we won't escape from these tabs in the future :)

**Comment by mtojek** [comment]
@ycombinator @jen-huang I've imported data streams definitions for AWS EC2 and DynamoDB

@ruflin I've added `spamfirewall` (Barracuda) data stream, which contains multiple agent templates (hbs).

**Comment by sorantis** [comment]
> @sorantis Clarifying question: based on the proposed package structure in this PR, which element in the package's `manifest.yml` would you like to see as separate tiles when a user searches for "aws" (besides the "main" AWS tile itself)?

I think each entry in `policy_templates` should be a tile. I.e. if the user searches for "aws", the search results should include all supported AWS services, even the ones from a different package, like AWS Fargate.

File: test/packages/policy_groups/manifest.yml
Comment (sorantis) [reply test/packages/policy_groups/manifest.yml:8]: In some cases it's a bigger set, i.e. DynamoDB is an `aws` service, a `cloud` service and a `datastore` service. `datastore` here is specific to DynamoDB and will not be applicable to other, for example compute related services.

**Comment by sorantis** [comment]
FYI, I updated the document to reflect the definitions we're using in the `manifest.yml` + it's more aligned with security. 
Changes: 
`integration instance` -> `integration policy`
`integration policy set` -> `integration policy template`
`integration policy` -> `input group` (or data group, or TBD. Basically it's the Logs/Metrics groups)

File: test/packages/policy_groups/img/aws-dynamodb.svg
Comment (ruflin) [reply test/packages/policy_groups/img/aws-dynamodb.svg]: 👍  This also has the advantage that the same icon can be reused ;-)

File: test/packages/policy_groups/manifest.yml
Comment (ruflin) [reply test/packages/policy_groups/manifest.yml]: The question is what happens if someone clicks on a category, what will the filter return:

* All integration sets + package
* Only package
* Only integration sets

Based on the above we should know if we need categories inside the integration sets / policy templates.

File: test/packages/input_groups/manifest.yml
Comment (ruflin) [inline test/packages/input_groups/manifest.yml:107]: Do we pull in here the additional vars from the data_stream manifest.yml? How do we know which one to select? What if there are multiple data_streams which configure `udp`? I remember I asked the question below but don't remember the answer :-(

Comment (mtojek) [reply test/packages/input_groups/manifest.yml:107]: > Do we pull in here the additional vars from the data_stream manifest.yml?

I think so.

Comment (mtojek) [reply test/packages/input_groups/manifest.yml:107]: > How do we know which one to select? What if there are multiple data_streams which configure udp? 

I don't reckon if we came up with a solution. Do you think we should introduce a "policy_templates" field to data stream manifest or do you see other options?

File: test/packages/policy_groups/manifest.yml
Comment (ycombinator) [reply test/packages/policy_groups/manifest.yml]: I'd prefer to start with the most restrictive option we can get away with and then make things less restrictive / more flexible depending on actual usage. So I'm ++ for hardcoding the group names (`logs` or `metrics`) for now. 

How far do we take this hardcoding? I assume not just the group `name`s are hardcoded but also the `title`s — "Logs" or "Metrics". What about the descriptions — should those be somewhat generated, e.g. "Logs for $package_name package"?

File: test/packages/input_groups/manifest.yml
Comment (ycombinator) [reply test/packages/input_groups/manifest.yml:107]: Yes, my understanding is also that such a problem (or is it a feature? 😎) already exists. I'm wary of changing this behavior, unless we can keep the _default_ behavior the same as it is today.

**Comment by ycombinator** [comment]
> I think each entry in `policy_templates` should be a tile. I.e. if the user searches for "aws", the search results should include all supported AWS services, even the ones from a different package, like AWS Fargate. 

So we will need to decide (most likely outside the scope of this PR) whether the package registry's search API response represents a list of packages (status quo) or a list of packages + policy templates.

Also, I'm not sure how to make this part work under the current proposal:

> ... even the ones from a different package, like AWS Fargate. 

Just to get the obvious question out of the way first: is there a reason AWS Fargate is it's own package and not a data stream under the `aws` package?

File: test/packages/policy_groups/manifest.yml
Comment (mtojek) [reply test/packages/policy_groups/manifest.yml]: Renamed to `input_groups`.

File: test/packages/policy_groups/manifest.yml
Comment (mtojek) [reply test/packages/policy_groups/manifest.yml]: @sorantis Any thoughts about hardcoding group names (logs, metrics, later on - checks)?

File: test/packages/input_groups/manifest.yml
Comment (mtojek) [reply test/packages/input_groups/manifest.yml:107]: What do you think about putting this is in a separate PR if this became an issue? We know that it can be solved with an additional field - `policy_templates` in `data_streams`.

File: test/packages/policy_groups/manifest.yml
Comment (mtojek) [reply test/packages/policy_groups/manifest.yml]: Yeah and extending to what @ruflin has said, it's more natural to keep DynamoDB, EC2 and others as separate packages ;)

**Comment by mtojek** [comment]
I think we should move towards the final decision about this pull request. Apart from:
* category vs subcategories for policy templates
* hardcoded input groups

do you see any other concerns around this PR?

**Comment by sorantis** [comment]
> Just to get the obvious question out of the way first: is there a reason AWS Fargate is it's own package and not a data stream under the `aws` package?

I guess it's due to the different configuration options and API endpoints, but @kaiyan-sheng will know more.

> I think we should move towards the final decision about this pull request. Apart from:
> 
> * category vs subcategories for policy templates
> * hardcoded input groups
> 
> do you see any other concerns around this PR?

I would avoid hardcoding input groups, one of the use cases we had in mind is to distinguish between operational and security logs (for setting different retentions). If we hardcode the input groups now, it will be hard for us to have this flexibility.

**Comment by mtojek** [comment]
> I would avoid hardcoding input groups, one of the use cases we had in mind is to distinguish between operational and security logs (for setting different retentions). If we hardcode the input groups now, it will be hard for us to have this flexibility.

I'm not sure if this PR assumes retention adjustments. It seems that input groups are more like UI tabs. Data retention is related rather to data stream configuration. Do we want to include retention settings in this PR or follow ups when necessary? I would vote for iterative approach, keep in mind that the first candidate is the AWS integration.

File: test/packages/policy_groups/manifest.yml
Comment (sorantis) [reply test/packages/policy_groups/manifest.yml]: If the user click on say "Datastore" then they should be able to see all integration that cover databases, data stores, etc. which also includes cloud services. Since tiles are (will be) generated from policy_templates then I think it's the option 3 in our list. 
This creates another question though. What happens if the user searches for AWS only, do we show AWS as a tile? The answer yes if we create a policy template (integration set) that is called AWS. 

This should work automatically for other integrations, for example NGINX. When the user searches for Nginx we show the Nginx tile because there's a policy template for it, we don't have to use package level information for it.
Screen Shot 2021-02-23 at 11 25 10

**Comment by sorantis** [comment]
@mtojek fair enough. In my mind input groups serve two purposes: 1) separate config options on tabs; 2) set different retentions at the input group level. The question is then how would the user set different retentions for operational vs. security logs, if it's not the input groups that define such separation? I think if don't hardcode input groups it will be easier for the UI to reuse them for setting retention policies.

**Comment by mtojek** [comment]
>  I think if don't hardcode input groups it will be easier for the UI to reuse them for setting retention policies.

I don't have a strong opinion on that, I suppose it's rather a UI design issue (@ruflin and @jen-huang can speak for Kibana). You brought here "operational logs" and "security logs", which means that it already violates (or extends?) the default split (logs-metrics-checks). If this is the approach we'd like to follow, let's stick to user defined input groups.

File: test/packages/input_groups/manifest.yml
Comment (ycombinator) [reply test/packages/input_groups/manifest.yml:107]: Yes, I'm ++ to taking this specific discussion to a follow up issue/PR.

**Comment by kaiyan-sheng** [comment]
> Just to get the obvious question out of the way first: is there a reason AWS Fargate is it's own package and not a data stream under the `aws` package?

Yep @sorantis is right. In order to collect metrics from `awsfargate`, you have to deploy metricbeat/agent inside the same task as the ECS containers you want to monitor. It does not require AWS credentials (access key, secret key).

**Comment by ycombinator** [comment]
> Yep @sorantis is right. In order to collect metrics from `awsfargate`, you have to deploy metricbeat/agent inside the same task as the ECS containers you want to monitor. It does not require AWS credentials (access key, secret key). 

Thanks @kaiyan-sheng.

Thinking about it some more, I think the proposal in this PR is unrelated to the _italicized_ part of the following use case:

> I think each entry in `policy_templates` should be a tile. I.e. if the user searches for "aws", the search results should include all supported AWS services, _even the ones from a different package, like AWS Fargate_. 

The proposal in this PR will cover the non-italicized part. 

As a follow up we can figure out how to make the Package Registry's search API return the `aws` package, each of its `policy_templates`, and other related packages like `AWS Fargate` when the user searches for the `aws` package. The solution might involve relaxing the matching strictness in the Package Registry's search API code OR introducing additional search keywords at the package / policy_templates level OR something else. Either way, I'm good with deferring this discussion to a follow up issue/PR. 

My main concern was that nothing being introduced in this PR would later get in the way of solving for the italicized part of the above use case, and I think we are good on that front.

File: test/packages/input_groups/manifest.yml
Comment (ruflin) [reply test/packages/input_groups/manifest.yml:107]: The problem does not exist today because we only have 1 policy template so we need all inputs. But with multiple policy templates, not all data_streams are needed for each one. So I think we need to solve it here.

Comment (mtojek) [reply test/packages/input_groups/manifest.yml:107]: I'm looking for propositions/ideas. I've mentioned one idea above.

Comment (ruflin) [reply test/packages/input_groups/manifest.yml:107]: I think both ways you proposed are feasible. I'm leaning in the direction of specifying it in the policy template. Let me explain why.

Lets assume I built a policy_template for a use case. Now I decide, I also want to allow users to configure data_stream x. All I need to do, is update my policy template and add an entry for it.

Having it in the policy template also makes it easy to understand, which data streams are used. I don't need to go to each data stream to check if the policy is mentioned in there.

Comment (mtojek) [reply test/packages/input_groups/manifest.yml:107]: I'm afraid it (_specifying it in the policy template_ ) would violate the open-closed principle, but we can turn a blind eye here. Let me try to introduce it.

Comment (ruflin) [reply test/packages/input_groups/manifest.yml:107]: Just a proposal, happy to discuss other solutions :-)

Comment (mtojek) [reply test/packages/input_groups/manifest.yml:112]: I added the `data_streams` property. Actually I'm fine with both solutions, I wouldn't say it's a concept hard to understand :)

Comment (ycombinator) [reply test/packages/input_groups/manifest.yml:112]: I'm good with the proposed solution with two comments:

1. **Default behavior**: not specifying `data_streams` in a policy template needs to be interpreted as if all data streams have been specified. This would guarantee backwards compatibility. I assume such default behavior logic will need to go in the Fleet UI code?

2. **Referential integrity**: We'll need a way to validate that data streams specified under `data_streams` actually exist in the package.

Neither comment above blocks this PR; just things to address in follow up PRs.

Comment (mtojek) [reply test/packages/input_groups/manifest.yml:112]: @ycombinator do you have any preference where `data_streams` proporty should be placed, either in `policy_template` or `input`? For reference: https://github.com/elastic/package-spec/commit/ed04c0f3a449872d3f281ea250c921b340a43668#commitcomment-47498856

File: test/packages/policy_groups/manifest.yml
Comment (sorantis) [reply test/packages/policy_groups/manifest.yml]: Speaking of how to express categories in the manifest file, I suggest we keep the top level categories section required, and introduce an optional categories section to a policy template.
This way each policy template will inherit top level categories and also has the option to add more policy template specific categories if needed.
Thoughts? @mtojek @ycombinator @ruflin

Comment (mtojek) [reply test/packages/policy_groups/manifest.yml]: In other words, categories on the root level (in manifest) would be common buckets valid for all policy templates, right?

Comment (sorantis) [reply test/packages/policy_groups/manifest.yml]: yes, with the option to extend this list for each policy template. I.e. for AWS common buckets could be `cloud`, `aws` and DynamoDB could optionally extend this list with `datastores` by defining it at the policy template level.

File: test/packages/policy_groups/manifest.yml
Comment (sorantis) [reply test/packages/policy_groups/manifest.yml]: responded here https://github.com/elastic/package-spec/pull/137#issuecomment-784100742

**Comment by mtojek** [comment]
We had an online discussion around the left issues and I think we came to the agreement:
* input groups stays as is
* policy template defines additional property `data_streams` to select particular data streams that are compatible with a policy template
* top categories are common, categories on the policy template level extend them

@ycombinator @ruflin please approve/comment on the PR

@jen-huang please double check if the PR is fine from Kibana dev perspective and also approve/comment here

**Review by ycombinator — APPROVED** [review]
LGTM. Really nice work iterating on this proposal and driving consensus!

I know this proposal will impact `kibana` and `package-registry` code at minimum.  I think it will also need a follow up PR in `package-spec` to enforce referential integrity between `data_streams` mentioned in policy template items and folders under `<package root>/data_stream/`. I don't think it affects `elastic-package` code (but I could be wrong about that). And, of course, eventually we'll want to introduce the enhancements enabled by this proposal to packages in the `integrations` repo.

So right before we merge this PR, could we create a checklist in https://github.com/elastic/package-spec/issues/132 that will capture each of the above tasks in an ordered fashion (topological sort on the dependency graph)? This will ensure that we can safely rollout this change and have a single place to track progress on the rollout. Some examples of such rollout checklists can be found in https://github.com/elastic/package-spec/issues/106 and https://github.com/elastic/package-spec/issues/17.

**Review by sorantis — APPROVED** [review]
Great to see this moving closer to the finish line! Thank you all for your involvement.

Agree with @ycombinator on creating a list of changes that need to be covered by related components. From the priority perspective, I'd like the focus be on Phase 1 described in the doc.

**Comment by mtojek** [comment]
@jen-huang could you please double check if these spec changes are fine for Kibana?

**Review by jen-huang — APPROVED** [review]
The changes look good to me from a Kibana perspective.

I would like to do more thorough validation by testing it with a local Kibana instance but that would need package registry changes so that the new fields are served through the package info endpoint. I saw that the [registry changes item](https://github.com/elastic/package-spec/issues/132#issuecomment-788007017) calls for checking with AWS package changes, but if it would be faster to get a draft PR up with a test package (even this nice `input_groups` package done here 😉), that would be super helpful.

And just to double check: we expect Kibana to be backwards compatible for packages without input groups too, right?

**Comment by mtojek** [comment]
> And just to double check: we expect Kibana to be backwards compatible for packages without input groups too, right?

Right, I would say this is an extension to current packages.

**Pull request merged**
